Added the troubleshooting steps to QDS Plan count (#7443)
MicrosoftDocs/sql-docs: docs/relational-databases/performance/best-practice-with-the-query-store.md

  ACTION(sqlos.system_thread_id,sqlos.task_address,sqlserver.database_id,sqlserver.database_name)),
  ADD EVENT qds.query_store_unloaded(
  ACTION(sqlos.system_thread_id,sqlos.task_address,sqlserver.database_id,sqlserver.database_name)),
  ADD EVENT sqlserver.query_store_failed_to_capture_query(
  ACTION(sqlos.system_thread_id,sqlos.task_address,sqlserver.database_id,sqlserver.database_name)),
  ADD EVENT sqlserver.query_store_failed_to_find_resource_group(
  ACTION(sqlos.system_thread_id,sqlos.task_address,sqlserver.database_id,sqlserver.database_name)),
  ADD EVENT sqlserver.query_store_failed_to_load_forced_plan(
  ACTION(sqlos.system_thread_id,sqlos.task_address,sqlserver.database_id,sqlserver.database_name)),
  ADD EVENT sqlserver.query_store_persist_on_shutdown_failed(
  ACTION(sqlos.system_thread_id,sqlos.task_address,sqlserver.database_id,sqlserver.database_name))
  ADD TARGET package0.event_file(SET filename=N'QueryStore',max_file_size=(100))
  WITH (MAX_MEMORY=4096 KB,EVENT_RETENTION_MODE=ALLOW_SINGLE_EVENT_LOSS,MAX_DISPATCH_LATENCY=30 SECONDS,MAX_EVENT_SIZE=0     KB,MEMORY_PARTITION_MODE=NONE,TRACK_CAUSALITY=OFF,STARTUP_STATE=OFF)
```

> You can find number of plans stored in Query Store using the below DMV’s.

Reviewer: ```suggestion
You can find the number of plans stored in Query Store using the below DMV’s.
```

> [!TIP]
> When using an Object-Relational Mapping (ORM) solution such as Entity Framework (EF), application queries like manual LINQ query trees or certain raw SQL queries may not be parameterized, which impacts plan re-use and the ability to track queries in the Query Store. For more information, see [EF Query caching and parameterization](/ef/core/performance/advanced-performance-topics) and [EF Raw SQL Queries](/ef/core/querying/raw-sql).

### Find non-parameterized queries in Query Store

You can find the number of plans stored in Query Store using the below query, using query store DMVs:

```sql
  SELECT count(plan_id) plan_count, query_hash,Txt.query_text_id, Txt.query_sql_text
  FROM sys.query_store_plan AS Pl
  INNER JOIN sys.query_store_query AS Qry
      ON Pl.query_id = Qry.query_id
  INNER JOIN sys.query_store_query_text AS Txt
      ON Qry.query_text_id = Txt.query_text_id
         group by query_hash,Txt.query_text_id, Txt.query_sql_text
         order by 1 desc

```

> Use the following sample extended event session to capture non-parameterized query usage in Query Store.

Reviewer 2: ```suggestion
Use the following sample extended event session to capture non-parameterized query usage in Query Store. On SQL Server, this extended event session creates an event_file target in the 'c:\xevents\' folder. For Azure SQL Managed Instance, specify an Azure Blob Storage location instead. For more information, see [XEvent event_file for Azure SQL Managed Instance](/azure/azure-sql/database/xevent-code-event-file#phase-2-transact-sql-code-that-uses-azure-storage-container). The event 'qds.query_store_db_diagnostics' is not available for Azure SQL Database. 
```

> [!TIP]
> When using an Object-Relational Mapping (ORM) solution such as Entity Framework (EF), application queries like manual LINQ query trees or certain raw SQL queries may not be parameterized, which impacts plan re-use and the ability to track queries in the Query Store. For more information, see [EF Query caching and parameterization](/ef/core/performance/advanced-performance-topics) and [EF Raw SQL Queries](/ef/core/querying/raw-sql).

### Find non-parameterized queries in Query Store

You can find the number of plans stored in Query Store using the below query, using query store DMVs, in [!INCLUDE[ssNoVersion](../../includes/ssnoversion-md.md)], [!INCLUDE[ssazuremi_md](../../includes/ssazuremi_md.md)], or [!INCLUDE[ssSDSfull](../../includes/sssdsfull-md.md)]:

```sql

>   SELECT count(Pl.plan_id) AS plan_count, Qry.query_hash, Txt.query_text_id, Txt.query_sql_text
>   FROM sys.query_store_plan AS Pl
>   INNER JOIN sys.query_store_query AS Qry
>       ON Pl.query_id = Qry.query_id
>   INNER JOIN sys.query_store_query_text AS Txt
>       ON Qry.query_text_id = Txt.query_text_id
>   GROUP BY Qry.query_hash, Txt.query_text_id, Txt.query_sql_text
>   ORDER BY plan_count desc;

Reviewer 2: ```suggestion
SELECT count(Pl.plan_id) AS plan_count, Qry.query_hash, Txt.query_text_id, Txt.query_sql_text
FROM sys.query_store_plan AS Pl
INNER JOIN sys.query_store_query AS Qry
    ON Pl.query_id = Qry.query_id
INNER JOIN sys.query_store_query_text AS Txt
    ON Qry.query_text_id = Txt.query_text_id
GROUP BY Qry.query_hash, Txt.query_text_id, Txt.query_sql_text
ORDER BY plan_count desc;
```

```sql
  SELECT count(Pl.plan_id) AS plan_count, Qry.query_hash, Txt.query_text_id, Txt.query_sql_text
  FROM sys.query_store_plan AS Pl
  INNER JOIN sys.query_store_query AS Qry
      ON Pl.query_id = Qry.query_id
  INNER JOIN sys.query_store_query_text AS Txt
      ON Qry.query_text_id = Txt.query_text_id
  GROUP BY Qry.query_hash, Txt.query_text_id, Txt.query_sql_text
  ORDER BY plan_count desc;
```

On SQL Server, this extended event session creates an event file in the SQL Server Log folder by default. For example, in a default SQL Server 2019 installation on Windows, the event file (.xel file) should be created in the folder `C:\Program Files\Microsoft SQL Server\MSSQL15.MSSQLSERVER\MSSQL\Log`. For Azure SQL Managed Instance, specify an Azure Blob Storage location instead. For more information, see [XEvent event_file for Azure SQL Managed Instance](/azure/azure-sql/database/xevent-code-event-file#phase-2-transact-sql-code-that-uses-azure-storage-container). The event 'qds.query_store_db_diagnostics' is not available for Azure SQL Database.  

The following sample creates a session to capture the event `query_store_db_diagnostics`:

```sql

>   CREATE EVENT SESSION [QueryStore_Troubleshoot] ON SERVER 
>   ADD EVENT qds.query_store_db_diagnostics(
>       ACTION(sqlos.system_thread_id,sqlos.task_address,sqlos.task_time,sqlserver.database_id,sqlserver.database_name))
>   ADD TARGET package0.event_file(SET filename=N'QueryStore',max_file_size=(100))
>   WITH (MAX_MEMORY=4096 KB,EVENT_RETENTION_MODE=ALLOW_SINGLE_EVENT_LOSS,MAX_DISPATCH_LATENCY=30 SECONDS,MAX_EVENT_SIZE=0 KB,MEMORY_PARTITION_MODE=NONE,TRACK_CAUSALITY=OFF,STARTUP_STATE=OFF)

Reviewer 2: ```suggestion
CREATE EVENT SESSION [QueryStore_Troubleshoot] ON SERVER 
ADD EVENT qds.query_store_db_diagnostics(
      ACTION(sqlos.system_thread_id,sqlos.task_address,sqlos.task_time,sqlserver.database_id,sqlserver.database_name))
ADD TARGET package0.event_file(SET filename=N'QueryStore',max_file_size=(100))
WITH (MAX_MEMORY=4096 KB,EVENT_RETENTION_MODE=ALLOW_SINGLE_EVENT_LOSS,MAX_DISPATCH_LATENCY=30 SECONDS,MAX_EVENT_SIZE=0 KB,MEMORY_PARTITION_MODE=NONE,TRACK_CAUSALITY=OFF,STARTUP_STATE=OFF);
```

```

With this data you can find plan count in the Query Store, and also many other stats as well. Look for the `plan_count`, `query_count`, `max_stmt_hash_map_size_kb`, and `max_size_mb` columns in the event data, in order to understand the amount of memory used and number of plans that are tracked by Query Store. If the plan count is higher than normal, it may indicate an increase in non-parameterized queries. Use the below Query Store DMVs query to review the parameterized queries and non-parameterized queries in the Query Store.

For parameterized queries:
```sql

> select query_id,qsqt.query_sql_text
>   from sys.query_store_query qsq, 
>   	 sys.query_store_query_text qsqt
>   where qsq.query_text_id= qsqt.query_text_id 
>   	  and (query_parameterization_type<>0 or query_sql_text like '%@%')

Reviewer 2: ```suggestion
SELECT qsq.query_id, qsqt.query_sql_text
FROM sys.query_store_query qsq INNER JOIN 
  	sys.query_store_query_text qsqt
ON qsq.query_text_id= qsqt.query_text_id 
WHERE query_parameterization_type<>0 or query_sql_text like '%@%';
```

```

For non-parameterized queries:

```sql

> select query_id,qsqt.query_sql_text
>   from sys.query_store_query qsq, 
>   	 sys.query_store_query_text qsqt
>   where qsq.query_text_id= qsqt.query_text_id 
>   	  and (query_parameterization_type=0)

Reviewer 2: ```suggestion
SELECT qsq.query_id, qsqt.query_sql_text
FROM sys.query_store_query AS qsq 
INNER JOIN sys.query_store_query_text AS qsqt
ON qsq.query_text_id= qsqt.query_text_id 
WHERE query_parameterization_type=0;
```

> [!TIP]
> When using an Object-Relational Mapping (ORM) solution such as Entity Framework (EF), application queries like manual LINQ query trees or certain raw SQL queries may not be parameterized, which impacts plan re-use and the ability to track queries in the Query Store. For more information, see [EF Query caching and parameterization](/ef/core/performance/advanced-performance-topics) and [EF Raw SQL Queries](/ef/core/querying/raw-sql).

### Find non-parameterized queries in Query Store

> You can find the number of plans stored in Query Store using the below query, using query store DMVs:

Reviewer 2: ```suggestion
You can find the number of plans stored in Query Store using the below query, using query store DMVs, in [!INCLUDE[ssNoVersion](../../includes/ssnoversion-md.md)], [!INCLUDE[ssazuremi_md](../../includes/ssazuremi_md.md)], or [!INCLUDE[ssSDSfull](../../includes/sssdsfull-md.md)]:
```

```sql
  SELECT count(plan_id) plan_count, query_hash,Txt.query_text_id, Txt.query_sql_text
  FROM sys.query_store_plan AS Pl
  INNER JOIN sys.query_store_query AS Qry
      ON Pl.query_id = Qry.query_id
  INNER JOIN sys.query_store_query_text AS Txt
      ON Qry.query_text_id = Txt.query_text_id
         group by query_hash,Txt.query_text_id, Txt.query_sql_text
         order by 1 desc

```

> On SQL Server, this extended event session creates an event file in the SQL Server Log folder by default. For example, in a default SQL Server 2019 installation on Windows, the event file (.xel file) should be created in the folder `C:\Program Files\Microsoft SQL Server\MSSQLSERVER\MSSQL\Log`. For Azure SQL Managed Instance, specify an Azure Blob Storage location instead. For more information, see [XEvent event_file for Azure SQL Managed Instance](/azure/azure-sql/database/xevent-code-event-file#phase-2-transact-sql-code-that-uses-azure-storage-container). The event 'qds.query_store_db_diagnostics' is not available for Azure SQL Database. 

Reviewer 2: ```suggestion
On SQL Server, this extended event session creates an event file in the SQL Server Log folder by default. For example, in a default SQL Server 2019 installation on Windows, the event file (.xel file) should be created in the folder `C:\Program Files\Microsoft SQL Server\MSSQL15.MSSQLSERVER\MSSQL\Log`. For Azure SQL Managed Instance, specify an Azure Blob Storage location instead. For more information, see [XEvent event_file for Azure SQL Managed Instance](/azure/azure-sql/database/xevent-code-event-file#phase-2-transact-sql-code-that-uses-azure-storage-container). The event 'qds.query_store_db_diagnostics' is not available for Azure SQL Database. 
```

> On SQL Server, this extended event session creates an event file in the SQL Server Log folder by default. For example, in a default SQL Server 2019 installation on Windows, the event file (.xel file) should be created in the folder `C:\Program Files\Microsoft SQL Server\MSSQL15.MSSQLSERVER\MSSQL\Log`. For Azure SQL Managed Instance, specify an Azure Blob Storage location instead. For more information, see [XEvent event_file for Azure SQL Managed Instance](/azure/azure-sql/database/xevent-code-event-file#phase-2-transact-sql-code-that-uses-azure-storage-container). The event 'qds.query_store_db_diagnostics' is not available for Azure SQL Database.  
> 
> The following sample creates a session to capture the event `query_store_db_diagnostics`:

Reviewer 2: ```suggestion
The following sample creates a session to capture the event `query_store_db_diagnostics`, which can be useful in diagnosing query resource consumption. In [!INCLUDE[ssNoVersion](../../includes/ssnoversion-md.md)], this extended event session creates an event file in the SQL Server Log folder by default. For example, in a default [!INCLUDE[sssql19-md](../../includes/sssql19-md.md)] installation on Windows, the event file (.xel file) should be created in the folder `C:\Program Files\Microsoft SQL Server\MSSQL15.MSSQLSERVER\MSSQL\Log`. For [!INCLUDE[ssazuremi_md](../../includes/ssazuremi_md.md)], specify an Azure Blob Storage location instead. For more information, see [XEvent event_file for Azure SQL Managed Instance](/azure/azure-sql/database/xevent-code-event-file#phase-2-transact-sql-code-that-uses-azure-storage-container). The event 'qds.query_store_db_diagnostics' is not available for [!INCLUDE[ssSDSfull](../../includes/sssdsfull-md.md)].  

```

```sql
  CREATE EVENT SESSION [QueryStore_Troubleshoot] ON SERVER 
  ADD EVENT qds.query_store_db_diagnostics(
      ACTION(sqlos.system_thread_id,sqlos.task_address,sqlos.task_time,sqlserver.database_id,sqlserver.database_name))
  ADD TARGET package0.event_file(SET filename=N'QueryStore',max_file_size=(100))
  WITH (MAX_MEMORY=4096 KB,EVENT_RETENTION_MODE=ALLOW_SINGLE_EVENT_LOSS,MAX_DISPATCH_LATENCY=30 SECONDS,MAX_EVENT_SIZE=0 KB,MEMORY_PARTITION_MODE=NONE,TRACK_CAUSALITY=OFF,STARTUP_STATE=OFF)
```

> With this data you can find plan count in the Query Store, and also many other stats as well. Look for the `plan_count`, `query_count`, `max_stmt_hash_map_size_kb`, and `max_size_mb` columns in the event data, in order to understand the amount of memory used and number of plans that are tracked by Query Store. If the plan_count is higher than normal, it may indicate an increase in non-parameterized queries. Use the below Query Store DMVs query to review the parameterized queries and non-parameterized queries in the Query Store.

Reviewer 2: ```suggestion
With this data you can find plan count in the Query Store, and also many other stats as well. Look for the `plan_count`, `query_count`, `max_stmt_hash_map_size_kb`, and `max_size_mb` columns in the event data, in order to understand the amount of memory used and number of plans that are tracked by Query Store. If the plan count is higher than normal, it may indicate an increase in non-parameterized queries. Use the below Query Store DMVs query to review the parameterized queries and non-parameterized queries in the Query Store.
```

Author: @WilliamDAssafMSFT, Good catch, Applied your suggested changes. Thanks!!

  ADD TARGET package0.event_file(SET filename=N'QueryStore',max_file_size=(100))
  WITH (MAX_MEMORY=4096 KB,EVENT_RETENTION_MODE=ALLOW_SINGLE_EVENT_LOSS,MAX_DISPATCH_LATENCY=30 SECONDS,MAX_EVENT_SIZE=0 KB,MEMORY_PARTITION_MODE=NONE,TRACK_CAUSALITY=OFF,STARTUP_STATE=OFF)
```

> The above extended event capture the event query_store_db_diagnostics, with this data you will find out the plan count in the Query Store and also many other stats as well.

Author: @PiJoCoder, I added 2 more queries for identifying the Parameterized Queries and Non-Parameterized Queries , As per the Jay , this extended event is help in case high volume of Non-Parameterized Queries it means high volume of plan count. So this extended event will indicate where is the root cause of performance issue.
![QueryStore](https://user-images.githubusercontent.com/62662636/162020389-cfd1e3c4-8922-4401-9c36-f67c1839b8aa.JPG)

> The above extended event capture the event query_store_db_diagnostics, with this data you will find out the plan count in the Query Store and also many other stats as well.

Author: I just tried the same at my local machine

> The above extended event capture the event query_store_db_diagnostics, with this data you will find out the plan count in the Query Store and also many other stats as well.

Reviewer: Yes, the added queries are great! Thanks for adding those - good learning experience for me. What William was asking is to tell the customer if they use the Xevent, what columns to look at - be explicit about it. That's why I am proposing this addition right here after this line.

```suggestion
The above extended event capture the event query_store_db_diagnostics, with this data you will find out the plan count in the Query Store and also many other stats as well. Look at the plan_count, query_count, max_stmt_hash_map_size_kb, max_size_mb columns in the event to understand the amount of memory used and number of plans that are tracked by Query Store.
```

> The above extended event capture the event query_store_db_diagnostics, with this data you will find out the plan count in the Query Store and also many other stats as well.

Author: Thanks @PiJoCoder  and @WilliamDAssafMSFT  for helping here. It will be a great learning for me as well. I have added the suggestions.

> The above extended event capture the event query_store_db_diagnostics, with this data you will find out the plan count in the Query Store and also many other stats as well.

Reviewer 2: I've made a couple commits to improve, please review. I think we're close! Any more direct info we can point to in the results of the `query_store_db_diagnostics` output?

> The above extended event capture the event query_store_db_diagnostics, with this data you will find out the plan count in the Query Store and also many other stats as well.

Author: @WilliamDAssafMSFT, I don't have handy to add more with respect to query_store_db_diagnostics output. Please merge it for now, if something comes, we will raise a new pull request.
Thank for your time and support on this, I appreciate it a lot.

> The above extended event capture the event query_store_db_diagnostics, with this data you will find out the plan count in the Query Store and also many other stats as well.

Reviewer 2: OK, made a few more style changes and edits to the code samples. I think we're ready to go now.
